## Adding more phonetic transcriptions


Out of the 2,510 distinct verbs in CHILDES uttered by children (0;0 to 7;0) and their caretakers, 552 were not found in the CELEX English verb IPA transcriptions (`english_merged.txt`). 

This python notebook saves the verbs in `verb_tokens_transcribed.csv` without IPA transcriptions into `verbs_not_in_CELEX.txt`.

We then feed the txt file to Claude for filtering, getting their past tense, transcribing the base and past tense to IPA, and getting their past tense type. 

The notebook also contains code for merging Claude's output to `verb_tokens_transcribed_full.csv`.

### Background
As a workaround, we tried creating transcriptions for these verbs using [NLTK's CMUdict](https://www.nltk.org/api/nltk.corpus.reader.cmudict.html), but ran into issues with getting the past tense of these verbs. We also had room to believe some of the tagged verbs were not actually verbs. (See the version history for this file for more information.)


Get list of verbs without a transcription from `verb_tokens_transcribed_ipa`

In [23]:
import pandas as pd

df = pd.read_csv('verb_tokens_transcribed.csv')

In [24]:
empty_df = df[df['stem_celex_encoding'].isnull()]
empty_df

,gloss,target_child_age,speaker_role,target_child_id,utterance_id,stem,stem_celex_encoding,past_tense,past_tense_celex_encoding
1,says,3.000062,Mother,2530,917970,say,NaN,NaN,NaN
9,fit,3.000062,Mother,2530,918404,fit,NaN,NaN,NaN
10,let's,3.000062,Mother,2530,918416,let,NaN,NaN,NaN
14,let's,3.000062,Mother,2530,918590,let,NaN,NaN,NaN
17,read,3.000062,Mother,2530,918647,read,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
1218416,do,35.986365,Target_Child,16213,10465001,do,NaN,NaN,NaN
1218417,has,35.986365,Target_Child,16213,10465026,have,NaN,NaN,NaN
1218419,do,35.986365,Target_Child,16213,10465128,do,NaN,NaN,NaN
1218420,have,35.986365,Target_Child,16213,10465171,have,NaN,NaN,NaN


In [25]:
verbs_to_add = empty_df['stem'].unique()

In [26]:
verbs_to_add

<StringArray>
[   'say',    'fit',    'let',   'read',     'do',    'put',   'make',
   'have',  'learn',    'dog',
 ...
  'armor',  'crest',  'tiger',   'molt', 'flurry',   'jive',  'scone',
  'hoove', 'lesson',   'sire']
Length: 552, dtype: str

For some reason, there seems to be some nouns in the missing verbs.

We utilize Anthropic's Claude to get the past tense and past tense type for the verbs not found in the CELEX transcription file. To do so, we save the verbs into a new `.txt` file.

See the directory [claude](./claude/) for more info on this.

In [32]:
verbs_to_add_file_path = "verbs_not_in_CELEX.txt"

with open(verbs_to_add_file_path, "w") as f:
    for word in verbs_to_add:
        f.write(f"{word}\n")

In [27]:
df['stem'].unique()

<StringArray>
[     'want',       'say',       'see',     'think',      'like',       'fit',
       'let',      'read',      'know',      'hold',
 ...
    'parade',     'hover',    'browse',       'ail',     'plonk',    'trudge',
      'spew', 'advertise',   'scuttle',      'sire']
Length: 2510, dtype: str

#### After getting Claude's output, we merge the info and save as a new file `verb_tokens_transcribed_full.csv`

All code henceforth coded with Claude

In [51]:
output_file = 'verb_tokens_transcribed_full.csv'

claude_output = './claude/verbs_with_past_tense.csv'
claude_filtered_out = './claude/excluded_non_verbs.txt'

Get the non-verbs that Claude filtered out and remove them from the verb_tokens_transcribed df

In [52]:
non_verbs = set()

with open(claude_filtered_out, 'r') as f:
    non_verbs = {line.strip() for line in f if line.strip()}

len(non_verbs)

203

In [53]:
df.shape

(1218430, 9)

In [54]:
filtered_df = df[~df['stem'].isin(non_verbs)]

filtered_df.shape

(1209501, 9)

Filtering out the 203 non-verbs using the `stem` column filtered out 8,929 rows.

In [55]:
# Filter out from CSV file with IPA transcriptions

ipa_transcribed = pd.read_csv('verb_tokens_transcribed_ipa.csv')
ipa_transcribed_filtered = ipa_transcribed[~ipa_transcribed['stem'].isin(non_verbs)]

In [56]:
ipa_transcribed_filtered.shape[0] == filtered_df.shape[0]

True

In [57]:
ipa_transcribed_filtered.shape

(1209501, 11)

In [58]:
claude_output_df = pd.read_csv(claude_output)

ipa_transcribed_filtered = ipa_transcribed_filtered.merge(
    claude_output_df[['verb', 'verb_ipa', 'past_tense', 'past_tense_ipa']],
    left_on='stem',
    right_on='verb',
    how='left'
)

ipa_transcribed_filtered['stem_ipa'] = ipa_transcribed_filtered['stem_ipa'].fillna(ipa_transcribed_filtered['verb_ipa'])
ipa_transcribed_filtered['past_tense'] = ipa_transcribed_filtered['past_tense_x'].fillna(ipa_transcribed_filtered['past_tense_y'])
ipa_transcribed_filtered['past_tense_ipa'] = ipa_transcribed_filtered['past_tense_ipa_x'].fillna(ipa_transcribed_filtered['past_tense_ipa_y'])

ipa_transcribed_filtered = ipa_transcribed_filtered.drop(columns=['verb', 'verb_ipa', 'past_tense_x', 'past_tense_y', 'past_tense_ipa_x', 'past_tense_ipa_y'])

In [59]:
ipa_transcribed_filtered.shape

(1209501, 11)

Drop `stem_celex_encoding`, `past_tense_celex_encoding` columns because they are not needed for modeling

In [60]:
ipa_transcribed_filtered = ipa_transcribed_filtered.drop(columns=['stem_celex_encoding', 'past_tense_celex_encoding'])

Make sure none of the IPA transcription columns are empty

In [61]:
ipa_transcribed_filtered[ipa_transcribed_filtered.isnull().any(axis=1)]

,gloss,target_child_age,speaker_role,target_child_id,utterance_id,stem,stem_ipa,past_tense,past_tense_ipa
770241,NaN,30.000616,Father,2622,1031798,screw,skɻuː:,screwed,skɻuː:d


Realized that the `verb_tokens_transcribed_ipa.csv` file did not the past tense type as a column.

### Add regular/irregular verb past tense type to the CSV output file

Adding the past tense type of Claude's output first

In [62]:
ipa_transcribed_filtered = ipa_transcribed_filtered.merge(
    claude_output_df[['verb', 'past_tense_type']],
    left_on='stem',
    right_on='verb',
    how='left'
).drop(columns='verb')

In [64]:
ipa_transcribed_filtered.columns

Index(['gloss', 'target_child_age', 'speaker_role', 'target_child_id',
       'utterance_id', 'stem', 'stem_ipa', 'past_tense', 'past_tense_ipa',
       'past_tense_type'],
      dtype='str')

Adding past tense type from CELEX file `english_merged.txt`

In [66]:
# Load the CELEX file
celex_df = pd.read_csv('english_merged.txt', sep='\t', header=None,
                       names=['stem', 'past_tense', 'stem_ipa', 'past_tense_ipa', 'past_tense_type'])

# Merge in the past_tense_type from CELEX, filling only where currently missing
celex_type = celex_df[['stem', 'past_tense_type']].rename(columns={'past_tense_type': 'past_tense_type_celex'})

ipa_transcribed_filtered = ipa_transcribed_filtered.merge(celex_type, on='stem', how='left')

ipa_transcribed_filtered['past_tense_type'] = ipa_transcribed_filtered['past_tense_type'].fillna(ipa_transcribed_filtered['past_tense_type_celex'])

ipa_transcribed_filtered = ipa_transcribed_filtered.drop(columns='past_tense_type_celex')

Make sure none of the IPA transcription or past tense type columns are empty

In [67]:
ipa_transcribed_filtered[ipa_transcribed_filtered.isnull().any(axis=1)]

,gloss,target_child_age,speaker_role,target_child_id,utterance_id,stem,stem_ipa,past_tense,past_tense_ipa,past_tense_type
770241,NaN,30.000616,Father,2622,1031798,screw,skɻuː:,screwed,skɻuː:d,reg


CELEX uses `reg` and `irreg` past tense types, while Claude used `regular` and `irregular`. We standardize them.

In [69]:
# Standardize to 'reg' and 'irreg'
ipa_transcribed_filtered['past_tense_type'] = ipa_transcribed_filtered['past_tense_type'].replace({
    'regular': 'reg',
    'irregular': 'irreg'
})

In [70]:
ipa_transcribed_filtered.to_csv(output_file, index=False)

In [71]:
ipa_transcribed_filtered.shape

(1209501, 10)